Support generation from input embedding

pfldy2850 opened the pull request:
This PR implements the feature of generating text from embedding input (popularly known as inputs_embeds).
This is related to https://github.com/vllm-project/vllm/issues/369 and https://github.com/vllm-project/vllm/issues/416.

More to do
- [x] Enhance test codes for generate.
- [x] Determine whether the feature reduces core performance.
- [x] Add more details to the comments.
- [x] apply it to async_llm_engine and api_server

pfldy2850 committed:
feat: add prompt_embeds interface

pfldy2850 committed:
fix: add get_input_embeddings

pfldy2850 committed:
feat: support all models to generate from embeds

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

pfldy2850 committed:
fix: bugfix for inputs_embeds and add last line

pfldy2850 committed:
fix: add prompt_embeds to async engine

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

pfldy2850 committed:
fix: bugfix of get_last_token_id

pfldy2850 committed:
fix: apply prompt_embeds to api_server

pfldy2850 committed:
refact: refactor test_models

pfldy2850 committed:
fix: apply style guide

pfldy2850 committed:
fix: improve comments

pfldy2850 committed:
refact: refactor prepare_inputs and models

pfldy2850 committed:
fix: apply style guide

pfldy2850 committed:
refact: refactor zero embeds

pfldy2850 commented:
We conducted several tests and confirmed that the performance degradation was not significant.

In fact, we measured the benchmark 5 times for the main branch and feature branch using the command below.

```
python benchmarks/benchmark_latency.py --input-len=2048 --num-iters=5

## main
Avg latency: 0.36247589644044637 seconds
Avg latency: 0.35677395705133674 seconds
Avg latency: 0.3622682703658938 seconds
Avg latency: 0.36043337155133487 seconds
Avg latency: 0.3593990854918957 seconds

## feature
Avg latency: 0.3586543008685112 seconds
Avg latency: 0.3557318979874253 seconds
Avg latency: 0.36645207908004523 seconds
Avg latency: 0.3598199490457773 seconds
Avg latency: 0.36111502479761837 seconds
```

pfldy2850 committed:
fix: apply style guide

chenxiaoliang1980 reviewed (approved):
fixed input embedding function related to #369 and #416

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

pfldy2850 committed:
fix: update for new prepare_inputs

pfldy2850 committed:
fix: rollback commented

pfldy2850 committed:
fix: update style

pfldy2850 commented:
@WoosukKwon @zhuohan123 

Hello authors, I have tested this PR and completed the alignment with the latest prepare_inputs function.
Could you please review this PR?

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

pfldy2850 committed:
fix: update model_runner with input_embeds

pfldy2850 committed:
fix: fix typo

pfldy2850 committed:
fix: bug fix

pfldy2850 committed:
fix: change input_embeds argument

pfldy2850 committed:
refact: refactor replace_prompt_embeds

pfldy2850 committed:
fix: bugfix

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

pfldy2850 commented:
Hi @WoosukKwon , 
I've made some changes to the PR that you saw, so I'm asking you to review it again.

- Updated source code to accept prompt_embeds argument in refactored code
- Updated to only handle prompt_embeds during _prepare_prompt (as it is unnecessary during the decoding part)
- Updated to accept prompt_embeds as an argument for all models (also added test code)
- Updated it available to async engine as well
- Updated entrypoints/api_server.py to accept prompt_embeds as a body

I have run the added test code and it passes all as shown below.
```
$ pytest tests/models/test_models.py -k test_models_from_prompt_embeds --forked
...
collected 22 items / 11 deselected / 11 selected                                                                                                                                         

tests/models/test_models.py ...........
```

I know you are very busy with a lot of interest and requests for vLLM. 
I would appreciate it if you could review this PR when you have a chance.

will-wiki commented:
@pfldy2850 
Thank you so much for your work! However, I encountered a problem when running this version of the [code](https://github.com/pfldy2850/vllm/tree/feature-input-embeds), almost the same environment, the same input and the same call method, the running result is not consistent with the official vLLM version.
The relevant call code, environment, and test results are as follows:

same call code：
`
import time
from vllm import LLM, SamplingParams
prompts = [
    '请介绍下爱因斯坦的生平。',
    '自由意志是否存在？',
]
sampling_params = SamplingParams(
    temperature=0, top_p=1, max_tokens=128, repetition_penalty=1.1,
    use_beam_search=True, best_of=5)

llm = LLM(model="internlm/internlm-7b", trust_remote_code=True)

outputs = []
start_time = time.perf_counter()
outputs = llm.generate(prompts, sampling_params, use_tqdm=False)
print(f"cost time: {(time.perf_counter() - start_time)/ len(prompts)} ")

# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
`

Operating environment V1：(office)
cuda：11.8
torch：2.1.0
xformers：0.0.22.post7+cu118
transformers：4.35.0
vllm：vllm @ https://github.com/vllm-project/vllm/releases/download/v0.2.2/vllm-0.2.2+cu118-cp39-cp39-manylinux1_x86_64.whl#sha256=7a8b51f0565baaa820f8dc0376e1ff5a732fcabda26397a55becd90e07b5fc63(office)

Result：
Prompt: '请介绍下爱因斯坦的生平。', Generated text: '\n爱因斯坦（1879年3月14日—1955 ），出生于德国巴登-符腾堡州乌尔姆市的一个犹太人家庭。他是20世纪最伟大的科学家之一、现代物理学的开创者和奠基人；他创立的狭义相对论和广义相…'
Prompt: '自由意志是否存在？', Generated text: '自由意志是否存在？\n首先，我们要明白什么是“决定论”。所谓**确定性理论（determinism）又称因果律、机械唯物主义和宿命说等：认为宇宙中一切事物的发展变化都有其必然的规律或原因可循；人的一切行为都由自然法则支配着并受物质运动规律的制约而发生和发展的一种哲学观点与世界观体系 【1】**\n其次要明确一个概念—意识是物质的反映还是独立于客观世界之外的存在呢?这个问题在科学界一直争论不休,至今没有定结论.但可以肯定的是:人的主观能动性的存在'

Operating environment V2：
cuda：11.8
torch：2.1.0
xformers：0.0.22.post7+cu118
transformers：4.35.0
vllm：https://github.com/pfldy2850/vllm/tree/feature-input-embeds（v0.2.3，install with 'pip install -e .'）

Result：
Prompt: '请介绍下爱因斯坦的生平。', Generated text: '-度小视\nPlease introduce the life of Einstein.\n1万播放 · 0弹幕2021年11月26日18：14更新·原创性地址信息新型科研机构研究基地享誉全球[s][s][]冰雹雷阵雨晴转多云东北风'
Prompt: '自由意志是否存在？', Generated text: ' - 哲学百科\n您现在的位置：首页 > TAG信息列表展示> Free Will Existence?\nFree will is the ability to make choices that are not determined by external factors. It has been debated for centuries, with some philosophers arguing it does exist and others claiming there\'s no such thing as free choice.\nThe debate over whether ornot we have a "free"will can be traced back at least two thousand years ago when Aristotle wrote about this topic in his book On The Soul (De Anima). He argued against determinism because he believed humans were capable of making decisions based on their own desires rather than being controlled'


The code has not changed, I am not clear whether the vllm installation is incorrect or there are other reasons, please help to look at it, thank you very much！if you need other information I can add.

pfldy2850 committed:
Merge branch 'main' into feature-input-embeds

js8544 commented:
We've been using this branch in production and it works like a charm. Thanks so much for your contribution. Can't wait for it to be merged!

fedshyvana commented:
thanks for this! Any plan to merge this into main anytime soon?

pfldy2850 commented:
Hello @zhuohan123 ,

I just saw that you created an issue for the vLLM Q1 2024 roadmap.

If you have any plans to consider this feature or merge for this PR,
I would like to resume the updating work for that PR.

matankley commented:
This PR would be super valuable for us. @pfldy2850  Do you plan to adjust it to the current master branch ? Because I see it is a bit outdated.

bks5881 commented on vllm/entrypoints/api_server.py:
```diff
@@ -29,16 +30,27 @@ async def generate(request: Request) -> Response:
     request_id = random_uuid()
 
-    results_generator = engine.generate(prompt, sampling_params, request_id)
+    results_generator = engine.generate(
```
I compared the speed with transformers library for prompt_embeds, its actually slower than transformers in multiple cases. Do you happen to know why is that?

bks5881 commented on vllm/entrypoints/api_server.py:
```diff
@@ -29,16 +30,27 @@ async def generate(request: Request) -> Response:
     prompt = request_dict.pop("prompt")
+    prompt_embeds = request_dict.pop("prompt_embeds", None)
+    if prompt_embeds is not None:
+        prompt_embeds = torch.tensor(prompt_embeds).to("cuda")
```
The loads stuff in float32. Eats all the GPU.

bks5881 commented on vllm/entrypoints/api_server.py:
```diff
@@ -29,16 +30,27 @@ async def generate(request: Request) -> Response:
     - other fields: the sampling parameters (See `SamplingParams` for details).
     """
     request_dict = await request.json()
     prompt = request_dict.pop("prompt")
```
This throws an error when only prompt_embeds are passed.

bks5881 reviewed (commented):
Thanks a lot for doing this. I tested it and had some issues I ran into but fixed them locally.
Also, for some reason when serializing  i got torch.cuda.is_available() as false. so had to set CUDA_VISBILE_DEVICES in ray. __init__.py

tweeter0830 commented:
@zhuohan123 Do you have plans for this? It would be really helpful to me for this MR to get merged. I can help push it through if you need.

zhuohan123 commented:
> @zhuohan123 Do you have plans for this? It would be really helpful to me for this MR to get merged. I can help push it through if you need.

We are doing this in this PR for llava support: #3042. Please take a look and let us know any suggestions!
